File: solution213.rs
impl Solution {
    pub fn robRange(nums: &Vec<i32>, start: usize, end: usize) -> i32 {
        let mut first: i32 = nums[start];
        let mut second: i32 = nums[start].max(nums[start+1]);

        for i in start+2..end {
            let mut temp = second;
            second = (nums[i] + first).max(second);
            first = temp;
        }
        return second;
    }

    pub fn rob(nums: Vec<i32>) -> i32 {
       let mut res1: i32 = 0;
       let mut res2: i32 = 0;
       
       if nums.len() == 1 {
           return nums[0];
       } else if nums.len() == 2 {
           return nums[0].max(nums[1]);
       }
       res1 = Solution::robRange(&nums, 0 as usize, nums.len()-1 as usize);
       res2 = Solution::robRange(&nums, 1 as usize, nums.len() as usize);

       return res1.max(res2);
    }
}

File: solution213.cpp
class Solution {
public:
    int robRange(vector<int>& nums, int start, int end) {
        int first = nums[start];
        int second = max(nums[start], nums[start + 1]);

        for (int i = start + 2; i <= end; i++) {
            int temp = second;
            second = max(nums[i] + first, second);
            first = temp;
        }

        return second;
    }

    int rob(vector<int>& nums) {
        int size = nums.size();
        
        if (size == 1) {
            return nums[0];
        }
        if (size == 2) {
            return max(nums[0], nums[1]);
        }

        return max(robRange(nums, 0, size - 2), robRange(nums, 1, size - 1));
    }
};
